src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -90,6 +92,19 @@ Chunk KafkaSource::generate()
     if (isCancelled())
         return {};
 
+    /// For non-streaming queries
+    if (high_watermark != Kafka::NO_WARTERMARK)
+    {
+        /// The topic is empty.
+        if (offset == high_watermark) [[unlikely]]
+            return {};
+
+        /// All available messages up to the moment when the query was executed have been consumed, no need to read the messages beyond that point.
+        /// `high_watermark` is the next available offset, i.e. the offset that will be assigned to the next message, thus need to use `high_watermark - 1`.
+        if (lastProcessedSN() >= high_watermark - 1)
```

Comment:
do we need this check ?

Reply:
Yes. That's the point to use `high_watermark`, so we know when to stop. The idea here is that, when `table()` is used on the external stream, we first find out what's the maximum offset, i.e. the high watermark, currently in the topic partition, and we make sure that we only read messages up to that offset.

For a less busy Kafka topic, without this check could be fine, but for a busy topic, without this check, it will never stop.